src/Compiler/Microsoft.AspNetCore.Razor.Language/src/DocumentationDescriptor.FormattedDescriptor.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Globalization;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+internal abstract partial class DocumentationDescriptor
+{
+    private sealed class FormattedDescriptor : DocumentationDescriptor
+    {
+        public override object?[] Args { get; }
+
+        private string? _formattedString;
+
+        public FormattedDescriptor(DocumentationId id, object?[] args)
+            : base(id)
+        {
+#if DEBUG
+            foreach (var arg in args)
+            {
+                Debug.Assert(
+                    arg is string || arg is int || arg is bool || arg is null,
+                    "Only string, int, bool, or null arguments are allowed.");
+            }
+#endif
+
+            Args = args;
+        }
+
+        public override string GetText()
+            => _formattedString ??= string.Format(CultureInfo.CurrentCulture, GetDocumentationText(), Args);
+
+        public override bool Equals(DocumentationDescriptor other)
+        {
+            if (other is not FormattedDescriptor { Id: var id, Args: var args })
+            {
+                return false;
+            }
+
+            if (Id != id)
+            {
+                return false;
+            }
+
+            var length = Args.Length;
+
+            if (length != args.Length)
+            {
+                return false;
+            }
+
+            for (var i = 0; i < length; i++)
+            {
+                var thisArg = Args[i];
+                var otherArg = args[i];
+
+                var areEqual = (thisArg, otherArg) switch
+                {
+                    (string s1, string s2) => s1 == s2,
+                    (int i1, int i2) => i1 == i2,
+                    (bool b1, bool b2) => b1 == b2,
+                    (object o1, object o2) => Equals(o1, o2),
```

Comment:
Is this case actually needed? We verified on input that only `string`, `int`, `bool`, and `null` are allowed: it seems like those should be handled by the rest of the cases in this switch?

Reply:
Yes, you're right. I think I had aspirations of allowing arbitrary object types and failed to remove this when I decided against that approach. I'll remove this unnecessary case.